React router

billbitt opened the pull request:
- added react-router
- rewrote relevant handlebars templates/partials as react components
- updated api routes to serve react

billbitt commented on react/components/AssetDisplay/index.js:
```diff
@@ -0,0 +1,133 @@
+    this.triggerGetAssetOnServer = this.triggerGetAssetOnServer.bind(this);
+  }
+  componentDidMount () {
+    const that = this;
```
I think I had a missunderstanding of how the `this` context works and when I needed to pass this in to a function manually.  I was able to remove it from the app in multiple places where it isn't necessary.

billbitt commented on react/components/AssetPreview/index.js:
```diff
@@ -0,0 +1,51 @@
+const AssetPreview = ({ name, claimId, fileExt, contentType }) => {
+  const directSourceLink = `${claimId}/${name}.${fileExt}`;
+  const showUrlLink = `${claimId}/${name}`;
+  const previewHolderStyle = {
```
oops

billbitt commented on react/containers/ChannelClaimsDisplay/index.js:
```diff
@@ -0,0 +1,27 @@
+
+const mapStateToProps = ({ show }) => {
+  return {
+    name       : show.showChannel.channelData.name,
```
They will always exist when ChannelClaimsDisplay is rendered, unless that should change... I am updated it to be destructured, is that what you were thinking?

seanyesmunt replied:
I was just wondering if there would ever be a case when `show` is undefined. Which would cause an errror. `cannot read property 'showChannel` of undefined`.

Or if any of those children would be undefined which would throw an error

billbitt commented on views/index.handlebars:
```diff
@@ -1,6 +1,5 @@
-    <div id="react-publish-tool" class="row row--padded row--tall flex-container--column">
+    <div id="react-app" class="row row--tall flex-container--column">
         <div class="row row--padded row--tall flex-container--column flex-container--center-center">
             <p>loading...</p>
```
I meant to keep the loading bar in so it will render until the app is rendered.  Once the app renders it replaces the content of the div and removes the loading bar.

billbitt commented on react/containers/PublishForm/view.jsx:
```diff
@@ -70,16 +71,14 @@ class PublishForm extends React.Component {
     xhr.open('POST', uri, true);
```
Cool.  Yeah, I replaced most xhr requests with fetch in the app.  I kept this one for the publish request because I wanted to set event listeners on the request and wasn't sure if I could do that with fetch.  I'll look into that.

billbitt commented on react/containers/NavBar/view.jsx:
```diff
@@ -63,17 +71,18 @@ class NavBar extends React.Component {
           <div className="nav-bar--right">
-            <a className="nav-bar-link link--nav-active" href="/">Publish</a>
-            <a className="nav-bar-link link--nav" href="/about">About</a>
+            <NavLink className="nav-bar-link link--nav" activeClassName="link--nav-active" to="/" exact={true}>Publish</NavLink>
```
I'm not quite sure I understand.  Do you mean create my own `NavLink` component ( rather than using the `NavLink` from `react-router-dom`) and control the className from inside the component?

seanyesmunt replied:
Oh I didn't realize that was a `react-router-dom` component. But yeah, I would make a custom `NavLink` component that uses the `react-router-dom` component. Then you can extend the functionality/style very easily.

seanyesmunt replied:
similar to this https://github.com/lbryio/lbry-app/blob/redesign-wip/src/renderer/component/link/view.jsx

billbitt commented on react/reducers/show.js:
```diff
@@ -0,0 +1,104 @@
+*/
+
+export default function (state = initialState, action) {
+  switch (action.type) {
```
Hmm, I like the readability of the switch statement, but I might use this util instead.  I have to look at the app and see exactly how it works.

seanyesmunt replied:
Here is an example of it in the app
https://github.com/lbryio/lbry-app/blob/master/src/renderer/redux/reducers/shape_shift.js#L99

billbitt commented on helpers/errorHandlers.js:
```diff
@@ -5,25 +5,11 @@ module.exports = {
     // check for daemon being turned off
     if (error.code === 'ECONNREFUSED') {
-      status = 503;
+      status = 200;
```
I am trying to create a pattern where server responded with a 200 as long as it processed the request correctly, and then handled the success of the data portion of the request with a `success` and a `message` or `data` field.  I.e. `{"success": "false", "message": "connection refused.  The daemon may not be running"}`.  This way, if the client wanted some data but the daemon was off, it would get a `200` response that the request was completed, and then the client could check the response to see if it was successful or not and whether the data is present. My goal was to help myself (and api users) to distinguish between an error where something was unavailable but the request was handled properly (resulting in a `200`) vs a request that I thought would be handled but wasn't, resulting in some other error.  Thoughts?

seanyesmunt replied:
I think proper error codes would still be preferable. Even if the request is handled properly, there is still an error. If I see a 200 response, I assume my request was successful. The `fetch` api handles errors really nicely too, so I think if anything you are creating more work.

```
fetch('spee.ch/xxx')
  .then((data) => ...)         // successful
  .catch((error) => ...)      // not successful

fetch('spee.ch/xxx')
  .then((data) => {           // maybe successful?
    if (!data.success) {
      throw new Error()
   }
    ...                                 // ok now I know it was successful
  })
  .catch((err) => ... )       // not successful
```
In my opinion it doesn't really matter if a request was handled successfully. If the data doesn't make it, it still fails. You can still add info with the error message and code.

seanyesmunt commented on react/actions/show.js:
```diff
@@ -0,0 +1,145 @@
+export function showNewChannel (name, shortId, longId) {
+  const id = `c#${name}#${longId}`;  // move to the action
+  return {
+    type: actions.CHANNEL_NEW_ASYNC,
```
This probably shouldn't be called `XXX_ASYNC` since it isn't async

seanyesmunt commented on react/containers/ShowAsset/view.jsx:
```diff
@@ -0,0 +1,50 @@
+  componentWillReceiveProps (nextProps) {
+    if (requestIsAnAssetRequest(nextProps)) {
+      const { asset, previousRequest, requestId, requestName, requestModifier } = nextProps;
+      if (!previousRequest) {
```
I think you are still creating more work than necessary with this. In my opinion `previousRequest` shouldn't even exist. In the `mapStateToProps` you should be able to map the `asset` from your state into the component. If `!asset` then make the request. 

I also think `onShowNewAsset` and `onNewRequest` can be combined. More specifically I don't think `onShowNewAsset` is needed. It might just be my lack of understanding with the current data flow, but you shouldn't need to manually say "show this asset". A better approach would be "select the asset with this id".

billbitt replied:
Ok, I think I'm getting closer.  I was able to do away with `onShowNewAsset` and combine the needed logic from its action (retrieving the asset's claim data) into `onNewRequest`.  That allowed me to remove `previousRequest` from the props I am passing to the `<ShowAsset />` component.  However, I am still checking for a `previousRequest` in the mapStateToProps function.  Do you see a way to avoid that step altogether?  The reason for storing and checking the previous requests is to avoid having to retrieve new information for a request that was already made (i.e. to avoid having to request the full `claimId` from the server).  I'm trying to figure out if that can be skipped or consolidated, but I am not sure how.

seanyesmunt commented on react/containers/ShowChannel/view.jsx:
```diff
@@ -0,0 +1,61 @@
+class ShowChannel extends React.Component {
+  componentDidMount () {
+    const { previousRequest, channel, requestId, requestChannelName, requestChannelId } = this.props;
+    if (!previousRequest) {
```
Same comments above about `previousRequest`. I think a more understandable approach would just be:
```
if (!channel) this.props.onNewChannelRequest(...)
```

billbitt replied:
See above re: previous request.  I mostly fixed this, but not sure if more consolidation can be done.

billbitt commented:
@seanyesmunt I was able to pull the request validation out of `<ShowAsset />` and remove that component entirely so that `<ShowPage />`  triggers the `onNewChannelRequest` or  `onNewAssetRequest` actions.  Those actions then retrieve the data and create the `request` along with the `asset` or `channel`.  I also consolidated the `channelRequests` and `assetRequests` into one `requestList`.  What do you think?

seanyesmunt commented:
@billbitt  Looks great. The components are definitely a lot simpler. I'm just now thinking that most of the logic inside `react/containers/ShowPage/view.jsx` can be moved into redux still. You can fetch the actual asset directly after fetching the longForm if you haven't seen that longForm before. No need to wait for the `ShowPage` to render to make the actual asset request.

Not a super big deal, but I think that should be done at some point. I started on some pseudo code and got a little carried away lol. But I think it could look something like this

```
// ShowPage.jsx
class ShowPage extends React.Component {
  componentDidMount() {
    this.props.handleShowPageUri(this.props.match.params)
  }

  render() { ... }
}

// Asset redux saga
function handleShowPageUri({ uri }) {
  if (!isValidUri(uri)) { // parse the uri just to make sure its valid
    dispatch(someError())
    return
  }

  // If this uri is in the request list, it's already been fetched
  if (state.requestsList[uri]) {
    return;
  }

  // Haven't seen this uri before
  return getLongForm(uri)
    .then((longForm) => {
      dispatch(addUriToRequestsList(uri))

      if (state.assetList[longForm]) {
        // we already have it, do nothing
        return
      } else {
        fetchAsset(uri)
          .then((asset) => {
            dispatch(addAsset(asset))
          })
      }
    })
}
```

billbitt commented:
@seanyesmunt I moved all that logic out of `<ShowPage />` and into a `handleShowPageUri` action.  That action parses it, validates it, and then triggers a `onNewChannelRequest` or `onNewAssetRequest` action.  In those actions, I added checks to the `store` to get the lists like in your pseudocode.  Can you check it out?

seanyesmunt commented on react/sagas/show_uri.js:
```diff
@@ -12,45 +13,46 @@ function parseAndUpdateIdentifierAndClaim (modifier, claim) {
   // return early if this request is for a channel
   if (isChannel) {
-    return onNewChannelRequest(channelName, channelClaimId);
+    return yield put(onNewChannelRequest(channelName, channelClaimId));
```
I think you need to add a state check here too. To see if that request exists. It looks like it will add the request to the list every time.
